persist/fs/files.go

```diff
@@ -96,53 +101,93 @@ func TimeFromFileName(fname string) (time.Time, error) {
 	return time.Unix(0, latestTimeInNano), nil
 }
 
-// InfoFiles returns all the completed info files in the given shard
-// directory, sorted by their corresponding block start times.
-func InfoFiles(shardDir string) ([]string, error) {
+// readCheckpointFile reads the checkpoint file and returns the digest stored
+// in the checkpoint file. Note that buf has digestSize bytes to avoid gc and
+// repeated memory allocation.
+func readCheckpointFile(shardDir string, blockStart time.Time, buf []byte) (uint32, error) {
+	checkpointFilePath := filepathFromTime(shardDir, blockStart, checkpointFileSuffix)
+	if !fileExists(checkpointFilePath) {
+		return 0, errCheckpointFileNotFound
+	}
+	checkpointFd, err := os.Open(checkpointFilePath)
+	if err != nil {
+		return 0, err
+	}
+	defer checkpointFd.Close()
+	return readDigestFromFile(checkpointFd, buf)
+}
+
+type infoFileFn func(fname string, infoData []byte)
+
+// ForEachInfoFile iterates over each valid info file and applies the function passed in.
+func ForEachInfoFile(filePathPrefix string, shard uint32, fn infoFileFn) {
+	shardDir := shardDirPath(filePathPrefix, shard)
 	matched, err := filepath.Glob(path.Join(shardDir, infoFilePattern))
 	if err != nil {
-		return nil, err
+		return
 	}
-	j := 0
+
+	sort.Sort(byTimeAscending(matched))
+	buf := make([]byte, digestLen)
 	for i := range matched {
 		t, err := TimeFromFileName(matched[i])
 		if err != nil {
-			return nil, err
+			continue
+		}
+		expectedDigestOfDigest, err := readCheckpointFile(shardDir, t, buf)
+		if err != nil {
+			continue
+		}
+		// Read and validate the digest file
+		digestData, err := readAndValidate(shardDir, t, digestFileSuffix, expectedDigestOfDigest)
+		if err != nil {
+			continue
 		}
-		checkpointFile := filepathFromTime(shardDir, t, checkpointFileSuffix)
-		if fileExists(checkpointFile) {
-			matched[j] = matched[i]
-			j++
+		// Read and validate the info file
+		expectedInfoDigest := readDigest(digestData[:digestLen])
+		infoData, err := readAndValidate(shardDir, t, infoFileSuffix, expectedInfoDigest)
+		if err != nil {
+			continue
 		}
+		fn(matched[i], infoData)
 	}
-	matched = matched[:j]
-	sort.Sort(byTimeAscending(matched))
-	return matched, nil
 }
 
-// FileExistsAt determines whether a data file exists for the given shard and block start time.
-func FileExistsAt(prefix string, shard uint32, blockStart time.Time) bool {
-	shardDir := ShardDirPath(prefix, shard)
-	checkpointFile := filepathFromTime(shardDir, blockStart, checkpointFileSuffix)
-	return fileExists(checkpointFile)
+// ReadInfoFiles reads all the valid info entries.
+func ReadInfoFiles(filePathPrefix string, shard uint32) []*schema.IndexInfo {
```

Comment:
Why do we want to read all the info files all at once?  This could potentially be hundreds of mbs.